sdc/rabbit/async_consumer.py

```diff
@@ -0,0 +1,388 @@
+        command. When it is complete, the on_queue_declareok method will
+        be invoked by pika.
+
+        :param str|unicode queue_name: The name of the queue to declare.
+
+        """
+        logger.info('Declaring queue', name=queue_name)
+        self._channel.queue_declare(
+            self.on_queue_declareok, queue_name, durable=self._durable_queue
+        )
+
+    def on_queue_declareok(self, method_frame):
+        """Method invoked by pika when the Queue.Declare RPC call made in
+        setup_queue has completed. In this method we will bind the queue
+        and exchange together with the routing key by issuing the Queue.Bind
+        RPC command. When this command is complete, the on_bindok method will
+        be invoked by pika.
+
+        :param pika.frame.Method method_frame: The Queue.DeclareOk frame
+
+        """
+        logger.info('Binding to rabbit', exchange=self._exchange, queue=self._queue)
+        self._channel.queue_bind(self.on_bindok, self._queue, self._exchange)
+
+    def add_on_cancel_callback(self):
+        """Add a callback that will be invoked if RabbitMQ cancels the consumer
+        for some reason. If RabbitMQ does cancel the consumer,
+        on_consumer_cancelled will be invoked by pika.
+
+        """
+        logger.info('Adding consumer cancellation callback')
+        self._channel.add_on_cancel_callback(self.on_consumer_cancelled)
+
+    def on_consumer_cancelled(self, method_frame):
+        """Invoked by pika when RabbitMQ sends a Basic.Cancel for a consumer
+        receiving messages.
+
+        :param pika.frame.Method method_frame: The Basic.Cancel frame
+
+        """
+        msg = 'Consumer was cancelled remotely, shutting down: {0!r]'
+        logger.info(msg.format(method_frame))
+        if self._channel:
+            self._channel.close()
+
+    def acknowledge_message(self, delivery_tag, **kwargs):
+        """Acknowledge the message delivery from RabbitMQ by sending a
+        Basic.Ack RPC method for the delivery tag.
+
+        :param int delivery_tag: The delivery tag from the Basic.Deliver frame
+
+        """
+        logger.info('Acknowledging message', delivery_tag=delivery_tag, **kwargs)
+        self._channel.basic_ack(delivery_tag)
+
+    def nack_message(self, delivery_tag, **kwargs):
+        """Negative acknowledge a message
+
+        :param int delivery_tag: The deliver tag from the Basic.Deliver frame
+
+        """
+        logger.info('Nacking message', delivery_tag=delivery_tag, **kwargs)
+        self._channel.basic_nack(delivery_tag)
+
+    def reject_message(self, delivery_tag, **kwargs):
+        """Reject the message delivery from RabbitMQ by sending a
+        Basic.Reject RPC method for the delivery tag.
+        :param int delivery_tag: The delivery tag from the Basic.Deliver frame
+
+        """
+        logger.info('Rejecting message', delivery_tag=delivery_tag, **kwargs)
+        self._channel.basic_reject(delivery_tag, requeue=False)
```

Comment:
I spoke to @necrophonic about this - the preference is to not requeue, but to quarantine in specific circumstances, otherwise we will enter a nacking cycle.